jib-cli/src/main/java/com/google/cloud/tools/jib/cli/cli2/Containerizers.java

```diff
@@ -0,0 +1,147 @@
+import com.google.cloud.tools.jib.api.ImageReference;
+import com.google.cloud.tools.jib.api.InvalidImageReferenceException;
+import com.google.cloud.tools.jib.api.LogEvent;
+import com.google.cloud.tools.jib.api.RegistryImage;
+import com.google.cloud.tools.jib.api.TarImage;
+import com.google.cloud.tools.jib.cli.cli2.logging.CliLogger;
+import com.google.cloud.tools.jib.event.events.ProgressEvent;
+import com.google.cloud.tools.jib.event.progress.ProgressEventHandler;
+import com.google.cloud.tools.jib.frontend.CredentialRetrieverFactory;
+import com.google.cloud.tools.jib.global.JibSystemProperties;
+import com.google.cloud.tools.jib.plugins.common.logging.ConsoleLogger;
+import com.google.cloud.tools.jib.plugins.common.logging.ProgressDisplayGenerator;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+
+/** Helper class for creating Containerizers from JibCli specifications. */
+public class Containerizers {
+
+  /**
+   * Create a Containerizer from a jibcli command line specification.
+   *
+   * @param buildOptions JibCli options
+   * @return a populated Containerizer
+   * @throws InvalidImageReferenceException if the image reference could not be parsed
+   */
+  public static Containerizer from(JibCli buildOptions) throws InvalidImageReferenceException {
+    ConsoleLogger logger =
+        CliLogger.newLogger(buildOptions.getVerbosity(), buildOptions.getConsoleOutput());
+
+    Containerizer containerizer = create(buildOptions, logger);
+
+    applyLogger(containerizer, logger);
+    applyConfiguration(containerizer, buildOptions);
+
+    return containerizer;
+  }
+
+  static Containerizer create(JibCli buildOptions, ConsoleLogger logger)
+      throws InvalidImageReferenceException {
+    String imageSpec = buildOptions.getTargetImage();
+    if (imageSpec.startsWith(DOCKER_DAEMON_IMAGE_PREFIX)) {
+      // TODO: allow setting docker env and docker executable (along with path/env)
+      return Containerizer.to(
+          DockerDaemonImage.named(imageSpec.replaceFirst(DOCKER_DAEMON_IMAGE_PREFIX, "")));
+    }
+    if (imageSpec.startsWith(TAR_IMAGE_PREFIX)) {
+      return Containerizer.to(
+          TarImage.at(Paths.get(imageSpec.replaceFirst(TAR_IMAGE_PREFIX, "")))
+              .named(buildOptions.getName()));
+    }
+    ImageReference imageReference =
+        ImageReference.parse(imageSpec.replaceFirst(REGISTRY_IMAGE_PREFIX, ""));
+    RegistryImage registryImage = RegistryImage.named(imageReference);
+    applyCredentialConfig(
+        registryImage,
+        CredentialRetrieverFactory.forImage(
+            imageReference, logEvent -> logger.log(logEvent.getLevel(), logEvent.getMessage())),
+        buildOptions);
+    return Containerizer.to(registryImage);
+  }
+
+  static void applyCredentialConfig(
+      RegistryImage registryImage, CredentialRetrieverFactory factory, JibCli buildOptions) {
+    if (buildOptions.getUsernamePassword().isPresent()) {
```

Comment:
It's not possible to have both `--username` and `--to-username` given, right? Is it worth checking that here?
```
Preconditions.checkArgument(!(buildOptions.getUsernamePassword().isPresent() && buildOptinos.getToUsernamePassword.isPresent()));
```

Reply:
From the cli, it's impossible. It could be a verify, but this is tested in the JibCli test.